File: mediainfo.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
)

var videoExts = map[string]bool{
	".mkv": true, ".mp4": true, ".avi": true, ".mov": true,
	".ts": true, ".m2ts": true, ".wmv": true, ".m4v": true,
}

// largestVideoFile returns the path of the largest video file under a directory.
func largestVideoFile(dir string) string {
	fi, err := os.Stat(dir)
	if err != nil || !fi.IsDir() {
		return dir
	}
	var best string
	var bestSize int64
	filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() {
			return nil
		}
		if videoExts[strings.ToLower(filepath.Ext(path))] && info.Size() > bestSize {
			bestSize = info.Size()
			best = path
		}
		return nil
	})
	return best
}

// GetMediaInfo extracts media info from a video file or the largest video in a directory.
// Uses the pure Go parser first (no dependencies).
// Falls back to ffprobe if the Go parser fails or the format is not supported.
func (a *App) GetMediaInfo(filePath string) (MediaInfo, error) {
	// If path is a directory, find the largest video file inside
	if fi, err := os.Stat(filePath); err == nil && fi.IsDir() {
		filePath = largestVideoFile(filePath)
		if filePath == "" {
			return MediaInfo{}, fmt.Errorf("aucun fichier vidéo trouvé dans le dossier")
		}
	}

	// Try pure Go parser first
	tracks, err := parseMediaFile(filePath)
	if err == nil && len(tracks) > 0 {
		info := tracksToMediaInfo(tracks, filePath)
		return info, nil
	}

	// Fall back to ffprobe
	info, ffErr := getMediaInfoFFprobe(filePath)
	if ffErr != nil {
		if err != nil {
			return MediaInfo{}, fmt.Errorf("go parser: %v | ffprobe: %v", err, ffErr)
		}
		return MediaInfo{}, fmt.Errorf("ffprobe non disponible (%v). Installez ffmpeg ou renseignez les infos manuellement.", ffErr)
	}
	return info, nil
}

// ─── ffprobe fallback ────────────────────────────────────────────────────────

type ffprobeOutput struct {
	Streams []ffprobeStream `json:"streams"`
	Format  ffprobeFormat   `json:"format"`
}

type ffprobeStream struct {
	CodecType      string            `json:"codec_type"`
	CodecName      string            `json:"codec_name"`
	Profile        string            `json:"profile"`
	Width          int               `json:"width"`
	Height         int               `json:"height"`
	RFrameRate     string            `json:"r_frame_rate"`
	ColorTransfer  string            `json:"color_transfer"`
	ColorPrimaries string            `json:"color_primaries"`
	Tags           map[string]string `json:"tags"`
	SideDataList   []ffprobeSideData `json:"side_data_list"`
	Channels       int               `json:"channels"`
}

type ffprobeSideData struct {
	SideDataType string `json:"side_data_type"`
}

type ffprobeFormat struct {
	Duration string `json:"duration"`
	Size     string `json:"size"`
	BitRate  string `json:"bit_rate"`
}

// ffprobePath returns the path to ffprobe: bundled next to the executable first, then PATH
func ffprobePath() string {
	exe, err := os.Executable()
	if err == nil {
		dir := filepath.Dir(exe)
		// macOS .app bundle: Contents/MacOS/ffprobe
		candidates := []string{
			filepath.Join(dir, "ffprobe"),
			filepath.Join(dir, "ffprobe.exe"),
		}
		for _, p := range candidates {
			if _, err := os.Stat(p); err == nil {
				return p
			}
		}
	}
	return "ffprobe" // fallback to PATH
}

func getMediaInfoFFprobe(filePath string) (MediaInfo, error) {
	cmd := exec.Command(ffprobePath(),
		"-v", "quiet",
		"-print_format", "json",
		"-show_streams",
		"-show_format",
		filePath,
	)
	out, err := cmd.Output()
	if err != nil {
		return MediaInfo{}, fmt.Errorf("ffprobe: %w", err)
	}
	var probe ffprobeOutput
	if err := json.Unmarshal(out, &probe); err != nil {
		return MediaInfo{}, fmt.Errorf("parse ffprobe output: %w", err)
	}
	return parseFFprobeOutput(probe, filePath), nil
}

func parseFFprobeOutput(probe ffprobeOutput, filePath string) MediaInfo {
	info := MediaInfo{}

	if probe.Format.Duration != "" {
		if d, err := strconv.ParseFloat(probe.Format.Duration, 64); err == nil {
			h := int(d) / 3600
			m := (int(d) % 3600) / 60
			s := int(d) % 60
			info.Duration = fmt.Sprintf("%02d:%02d:%02d", h, m, s)
		}
	}
	if probe.Format.Size != "" {
		if sz, err := strconv.ParseInt(probe.Format.Size, 10, 64); err == nil {
			info.FileSize = sz
		}
	}
	if probe.Format.BitRate != "" {
		if br, err := strconv.ParseInt(probe.Format.BitRate, 10, 64); err == nil {
			info.Bitrate = br
		}
	}

	var audioStreams []ffprobeStream
	for _, s := range probe.Streams {
		switch s.CodecType {
		case "video":
			if info.Width == 0 {
				info.Width = s.Width
				info.Height = s.Height
				info.Resolution = detectResolution(s.Width, s.Height)
				info.VideoCodec = normalizeVideoCodec(s.CodecName, s.Profile)
				info.HDRFormat = detectHDRFFprobe(s)
				if s.RFrameRate != "" {
					info.FrameRate = parseFrameRate(s.RFrameRate)
				}
			}
		case "audio":
			audioStreams = append(audioStreams, s)
		}
	}

	if len(audioStreams) > 0 {
		info.AudioCodec = bestFFprobeAudioCodec(audioStreams)
		info.AudioLanguages = collectFFprobeLanguages(audioStreams)
	}
	info.Source = guessSource(filePath)
	return info
}

// ─── shared helpers ──────────────────────────────────────────────────────────

func detectResolution(w, h int) string {
	switch {
	case h >= 2160 || w >= 3840:
		return "2160p"
	case h >= 1080 || w >= 1920:
		return "1080p"
	case h >= 720 || w >= 1280:
		return "720p"
	default:
		if w > 0 || h > 0 {
			return "SD"
		}
		return ""
	}
}

func normalizeVideoCodec(codec, profile string) string {
	switch strings.ToLower(codec) {
	case "hevc", "h265":
		return "H.265"
	case "h264", "avc":
		return "H.264"
	case "av1":
		return "AV1"
	case "vp9":
		return "VP9"
	case "mpeg4", "xvid", "divx":
		return "XviD"
	default:
		return strings.ToUpper(codec)
	}
}

func detectHDRFFprobe(s ffprobeStream) string {
	ct := strings.ToLower(s.ColorTransfer)
	isDV := false
	isHDR10Plus := false
	for _, sd := range s.SideDataList {
		sdl := strings.ToLower(sd.SideDataType)
		if strings.Contains(sdl, "dolby") {
			isDV = true
		}
		if strings.Contains(sdl, "hdr10+") || strings.Contains(sdl, "hdr10 plus") {
			isHDR10Plus = true
		}
	}
	if isDV && (ct == "smpte2084" || isHDR10Plus) {
		return "HDR DV"
	}
	if isDV {
		return "DV"
	}
	if isHDR10Plus {
		return "HDR10+"
	}
	if ct == "smpte2084" {
		return "HDR10"
	}
	if ct == "arib-std-b67" {
		return "HDR"
	}
	return ""
}

func bestFFprobeAudioCodec(streams []ffprobeStream) string {
	priority := map[string]int{
		"truehd": 10, "eac3": 8, "dts": 7, "flac": 6,
		"ac3": 5, "aac": 4, "mp3": 3,
	}
	best := ""
	bestPrio := -1
	for _, s := range streams {
		codec := strings.ToLower(s.CodecName)
		prio := priority[codec]
		if prio > bestPrio {
			bestPrio = prio
			best = s.CodecName
			if codec == "truehd" && s.Channels >= 8 {
				best = "truehd_atmos"
			}
			if codec == "dts" {
				switch strings.ToUpper(s.Profile) {
				case "DTS-HD MA":
					best = "dts_hd_ma"
				case "DTS-HD":
					best = "dts_hd"
				}
			}
		}
	}
	return normalizeAudioCodec(best)
}

func normalizeAudioCodec(codec string) string {
	switch strings.ToLower(codec) {
	case "truehd_atmos":
		return "Atmos"
	case "truehd":
		return "TrueHD"
	case "dts_hd_ma":
		return "DTS-HD MA"
	case "dts_hd":
		return "DTS-HD"
	case "dts":
		return "DTS"
	case "eac3":
		return "EAC3"
	case "ac3":
		return "AC3"
	case "flac":
		return "FLAC"
	case "aac":
		return "AAC"
	case "mp3":
		return "MP3"
	default:
		return strings.ToUpper(codec)
	}
}

var langNames = map[string]string{
	"fre": "Français (France)", "fra": "Français (France)", "fr": "Français (France)",
	"eng": "Anglais", "en": "Anglais",
	"spa": "Espagnol", "es": "Espagnol",
	"ger": "Allemand", "deu": "Allemand", "de": "Allemand",
	"ita": "Italien", "it": "Italien",
	"por": "Portugais", "pt": "Portugais",
	"jpn": "Japonais", "ja": "Japonais",
	"chi": "Chinois", "zho": "Chinois", "zh": "Chinois",
	"kor": "Coréen", "ko": "Coréen",
	"ara": "Arabe", "ar": "Arabe",
	"rus": "Russe", "ru": "Russe",
	"und": "",
}

func collectFFprobeLanguages(streams []ffprobeStream) string {
	seen := map[string]bool{}
	var langs []string
	for _, s := range streams {
		lang := ""
		if s.Tags != nil {
			lang = s.Tags["language"]
		}
		name := normalizeLangCode(lang)
		if name != "" && !seen[name] {
			seen[name] = true
			langs = append(langs, name)
		}
	}
	return strings.Join(langs, ", ")
}

func guessSource(filePath string) string {
	name := strings.ToLower(filepath.Base(filePath))
	switch {
	case strings.Contains(name, "bluray") || strings.Contains(name, "blu-ray") ||
		strings.Contains(name, "bdremux") || strings.Contains(name, "bdrip"):
		return "BluRay"
	case strings.Contains(name, "web-dl") || strings.Contains(name, "webdl"):
		return "WEB-DL"
	case strings.Contains(name, "webrip"):
		return "WEBRip"
	case strings.Contains(name, "hdtv"):
		return "HDTV"
	case strings.Contains(name, "dvdrip") || strings.Contains(name, "dvd"):
		return "DVDRip"
	case strings.Contains(name, "dcp"):
		return "DCP"
	}
	return ""
}

func parseFrameRate(s string) float64 {
	parts := strings.Split(s, "/")
	if len(parts) != 2 {
		return 0
	}
	num, err1 := strconv.ParseFloat(parts[0], 64)
	den, err2 := strconv.ParseFloat(parts[1], 64)
	if err1 != nil || err2 != nil || den == 0 {
		return 0
	}
	return math.Round(num/den*100) / 100
}

File: frontend/src/services/mediainfo.ts
import MediaInfoFactory from 'mediainfo.js'
import { GetFileSize, ReadFileChunk } from '../../wailsjs/go/main/App'

export interface ParsedMediaInfo {
  resolution: string
  videoCodec: string
  audioCodec: string
  audioLanguages: string
  hdrFormat: string
  source: string
  duration: string
  fileSize: number
  width: number
  height: number
  bitrate: number
  frameRate: number
}

const CHUNK_SIZE = 256 * 1024 // 256 KB

async function analyzeFile(filePath: string): Promise<any> {
  const fileSize = await GetFileSize(filePath)

  const mi = await MediaInfoFactory({ format: 'object', locateFile: () => 'mediainfo.wasm' })

  const readChunk = async (size: number, offset: number): Promise<Uint8Array> => {
    const b64 = await ReadFileChunk(filePath, offset, size)
    const bin = atob(b64)
    const arr = new Uint8Array(bin.length)
    for (let i = 0; i < bin.length; i++) arr[i] = bin.charCodeAt(i)
    return arr
  }

  const result = await mi.analyzeData(() => fileSize, readChunk)
  mi.close()
  return result
}

function pickBestAudioCodec(tracks: any[]): string {
  const priority: Record<string, number> = {
    'TrueHD Atmos': 10, 'TrueHD': 9,
    'DTS-HD MA': 8, 'DTS-HD': 7, 'DTS': 6,
    'E-AC-3': 5, 'AC-3': 4, 'FLAC': 3, 'AAC': 2, 'MP3': 1,
  }
  let best = ''
  let bestPrio = -1
  for (const t of tracks) {
    const codec = t.Format || ''
    const addl = t.Format_AdditionalFeatures || ''
    let name = codec
    if (codec === 'TrueHD' && addl.includes('Atmos')) name = 'TrueHD Atmos'
    else if (codec === 'MLP FBA' && addl.includes('Atmos')) name = 'TrueHD Atmos'
    else if (codec === 'DTS' && t.Format_Profile?.includes('MA')) name = 'DTS-HD MA'
    else if (codec === 'DTS' && t.Format_Profile?.includes('HRA')) name = 'DTS-HD'
    const prio = priority[name] ?? 0
    if (prio > bestPrio) { bestPrio = prio; best = name }
  }
  return normalizeAudioCodec(best)
}

function normalizeAudioCodec(codec: string): string {
  const map: Record<string, string> = {
    'TrueHD Atmos': 'Atmos', 'TrueHD': 'TrueHD',
    'DTS-HD MA': 'DTS-HD MA', 'DTS-HD': 'DTS-HD', 'DTS': 'DTS',
    'E-AC-3': 'EAC3', 'AC-3': 'AC3', 'FLAC': 'FLAC', 'AAC': 'AAC', 'MP3': 'MP3',
  }
  return map[codec] ?? codec
}

const langNames: Record<string, string> = {
  fr: 'Français', fre: 'Français', fra: 'Français',
  en: 'Anglais', eng: 'Anglais',
  de: 'Allemand', ger: 'Allemand', deu: 'Allemand',
  es: 'Espagnol', spa: 'Espagnol',
  it: 'Italien', ita: 'Italien',
  ja: 'Japonais', jpn: 'Japonais',
  pt: 'Portugais', por: 'Portugais',
  zh: 'Chinois', chi: 'Chinois', zho: 'Chinois',
  ko: 'Coréen', kor: 'Coréen',
  ru: 'Russe', rus: 'Russe',
  ar: 'Arabe', ara: 'Arabe',
}

function normalizeLang(lang: string): string {
  return langNames[lang?.toLowerCase()] ?? lang ?? ''
}

function detectHDR(videoTrack: any): string {
  const transfer = videoTrack.transfer_characteristics_Original || videoTrack.transfer_characteristics || ''
  const hdrFormat = videoTrack.HDR_Format || ''
  const hdrCompat = videoTrack.HDR_Format_Compatibility || ''

  const isDV = hdrFormat.toLowerCase().includes('dolby vision')
  const isHDR10Plus = hdrFormat.toLowerCase().includes('hdr10+') || hdrCompat.toLowerCase().includes('hdr10+')
  const isHDR10 = transfer.includes('PQ') || transfer.includes('SMPTE ST 2084')
  const isHLG = transfer.includes('HLG') || transfer.includes('ARIB STD B67')

  if (isDV && isHDR10) return 'HDR DV'
  if (isDV) return 'DV'
  if (isHDR10Plus) return 'HDR10+'
  if (isHDR10) return 'HDR10'
  if (isHLG) return 'HDR'
  return ''
}

function normalizeVideoCodec(format: string, profile: string): string {
  const f = format?.toLowerCase() ?? ''
  if (f.includes('hevc') || f === 'h.265') return 'H.265'
  if (f.includes('avc') || f === 'h.264') return 'H.264'
  if (f === 'av1') return 'AV1'
  if (f === 'vp9') return 'VP9'
  if (f.includes('xvid') || f.includes('divx') || f.includes('mpeg-4')) return 'XviD'
  return format ?? ''
}

function detectResolution(w: number, h: number): string {
  if (h >= 2160 || w >= 3840) return '2160p'
  if (h >= 1080 || w >= 1920) return '1080p'
  if (h >= 720 || w >= 1280) return '720p'
  if (w > 0 || h > 0) return 'SD'
  return ''
}

function guessSource(filePath: string): string {
  const name = filePath.toLowerCase()
  if (name.includes('bluray') || name.includes('blu-ray') || name.includes('bdremux') || name.includes('bdrip')) return 'BluRay'
  if (name.includes('web-dl') || name.includes('webdl')) return 'WEB-DL'
  if (name.includes('webrip')) return 'WEBRip'
  if (name.includes('hdtv')) return 'HDTV'
  if (name.includes('dvdrip') || name.includes('dvd')) return 'DVDRip'
  if (name.includes('dcp')) return 'DCP'
  return ''
}

export async function getMediaInfoJS(filePath: string): Promise<ParsedMediaInfo> {
  const result = await analyzeFile(filePath)
  const tracks: any[] = result?.media?.track ?? []

  const general = tracks.find((t: any) => t['@type'] === 'General') ?? {}
  const video = tracks.find((t: any) => t['@type'] === 'Video') ?? {}
  const audioTracks = tracks.filter((t: any) => t['@type'] === 'Audio')

  const w = parseInt(video.Width ?? '0')
  const h = parseInt(video.Height ?? '0')

  // Duration
  let duration = ''
  const dur = parseFloat(general.Duration ?? '0')
  if (dur > 0) {
    const h_ = Math.floor(dur / 3600)
    const m = Math.floor((dur % 3600) / 60)
    const s = Math.floor(dur % 60)
    duration = `${String(h_).padStart(2, '0')}:${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`
  }

  // Audio languages
  const seen = new Set<string>()
  const langs: string[] = []
  for (const t of audioTracks) {
    const lang = normalizeLang(t.Language ?? '')
    if (lang && !seen.has(lang)) { seen.add(lang); langs.push(lang) }
  }

  // Frame rate
  const fps = parseFloat(video.FrameRate ?? '0')

  return {
    resolution: detectResolution(w, h),
    videoCodec: normalizeVideoCodec(video.Format ?? '', video.Format_Profile ?? ''),
    audioCodec: pickBestAudioCodec(audioTracks),
    audioLanguages: langs.join(', '),
    hdrFormat: detectHDR(video),
    source: guessSource(filePath),
    duration,
    fileSize: parseInt(general.FileSize ?? '0'),
    width: w,
    height: h,
    bitrate: parseInt(general.OverallBitRate ?? '0'),
    frameRate: Math.round(fps * 100) / 100,
  }
}
